File: practice/segtree.cpp
#include <bits/stdc++.h>

using namespace std;

template<typename T> class SegTree {
    // Segment tree가 구하고자 하는 값에 따라 변경
    const T BOUND = 0;
    T f(T a, T b) { return a + b; }

    int MIN, MAX;
    vector<T> node;

    T query(int k, int node_s, int node_e, int req_s, int req_e) {
        if (req_e < node_s || node_e < req_s)
            return BOUND;
        if (req_s <= node_s && node_e <= req_e)
            return node[k];
        int node_m = (node_s + node_e) / 2;
        T left = query(k * 2, node_s, node_m, req_s, req_e);
        T right = query(k * 2 + 1, node_m + 1, node_e, req_s, req_e);
        return f(left, right);
    }
    void update(int k, int node_s, int node_e, int req_i, T value) {
        if (req_i < node_s || node_e < req_i)
            return;
        if (node_s == node_e) {
            node[k] = value;
            return;
        }
        int node_m = (node_s + node_e) / 2;
        update(k * 2, node_s, node_m, req_i, value);
        update(k * 2 + 1, node_m + 1, node_e, req_i, value);
        node[k] = f(node[k * 2], node[k * 2 + 1]);
    }
public:
    SegTree(int MIN, int MAX) : MIN(MIN), MAX(MAX) { node.resize(4 * (MAX - MIN)); }
    T query(int start, int end) { return query(1, MIN, MAX, start, end); }
    void update(int i, T value) { update(1, MIN, MAX, i, value); }
};

int main() {
    vector<int> v = {1, 2, 3, 4, 5, 6, 7};
    SegTree<int> st(0, 6);
    for(int i=0; i<(int)v.size(); i++) {
        st.update(i, v[i]);
    }
    cout << st.query(0, 3) << endl;
    cout << st.query(3, 6) << endl;
    cout << st.query(2, 5) << endl;
}

File: algorithms/seg_tree.py
class SegTree:
    def __init__(self, data):
        self.data = data
        self.tree = [0] * (4 * len(self.data))
        self.init(1, 0, len(self.data) - 1)

    def init(self, ti, l, r):
        if l == r:
            self.tree[ti] = self.data[l]
            return self.tree[ti]
        self.tree[ti] = self.init(ti * 2, l, (l + r) // 2) \
            + self.init(ti * 2 + 1, (l + r) // 2 + 1, r)
        return self.tree[ti]

    def __update__(self, ti, di, elt, l, r):
        if di < l or r < di:
            return self.tree[ti]
        if l == r:
            self.tree[ti] = elt
            return self.tree[ti]
        self.tree[ti] = self.__update__(ti * 2, di, elt, l, (l + r) // 2) \
            + self.__update__(ti * 2 + 1, di, elt, (l + r) // 2 + 1, r)
        return self.tree[ti]

    def __query__(self, ti, ql, qr, tl, tr):
        if ql <= tl and tr <= qr:
            return self.tree[ti]
        if tr < ql or qr < tl:
            return 0
        return self.__query__(ti * 2, ql, qr, tl, (tl + tr) // 2) \
            + self.__query__(ti * 2 + 1, ql, qr, (tl + tr) // 2 + 1, tr)

    def update(self, idx, elt):
        self.__update__(1, idx, elt, 0, len(self.data) - 1)

    def query(self, left, right):
        return self.__query__(1, left, right, 0, len(self.data) - 1)


def main():
    st = SegTree([1, 2, 3, 4, 5, 6, 7, 8, 9])

    print(st.query(4, 8))
    st.update(2, 1)
    st.update(5, 3)
    st.update(6, 10)
    print(st.query(2, 3))
    print(st.query(0, 6))


if __name__ == '__main__':
    main()
